lib-ce/emqx_modules/src/emqx_mod_rewrite.erl

```diff
@@ -75,30 +74,60 @@ description() ->
 %% Internal functions
 %%--------------------------------------------------------------------
 
-log_start(Rules) ->
-    PubRules = [{pub, Topic, Re, Dest} || {rewrite, pub, Topic, Re, Dest} <- Rules],
-    SubRules = [{sub, Topic, Re, Dest} || {rewrite, sub, Topic, Re, Dest} <- Rules],
+compile_rules(RawRules) ->
+    compile(validate_rules(RawRules)).
+
+compile({PubRules, SubRules}) ->
+    CompileRE =
+        fun({rewrite, Source, Re, Dest}) ->
+                {ok, MP} = re:compile(Re),
+                {rewrite, Source, MP, Dest}
+        end,
+    {lists:map(CompileRE, PubRules), lists:map(CompileRE, SubRules)}.
+
+validate_rules(Rules) ->
+    validate_rules(Rules, {[], []}).
+
+validate_rules([], {PubRules, SubRules}) ->
     ?LOG(info, "[Rewrite] Load: pub rules count ~p sub rules count ~p",
         [erlang:length(PubRules), erlang:length(SubRules)]),
-    log_rule(PubRules, 1),
-    log_rule(SubRules, 1).
+    log_rules(pub, PubRules),
+    log_rules(sub, SubRules),
+    {PubRules, SubRules};
+
+validate_rules([Rule = {_, Type, _, _, _} | RestRules], {PubRules, SubRules}) ->
+    case {Type, validate_rule(Rule)} of
+        {pub, {ok, NRule}} ->
+            validate_rules(RestRules, {[NRule| PubRules], SubRules});
+        {sub, {ok, NRule}} ->
+            validate_rules(RestRules, {PubRules, [NRule | SubRules]});
+        {_, {error, {Direction, T}, Reason}} ->
+            ?LOG(warning, "Topic rewrite ~p rule item validate error. The ~p topic ~p invalid, Reason: ~p",
+                 [assert_type(Type), Direction, T, Reason]),
+            validate_rules(RestRules, {PubRules, SubRules})
+    end.
+
+validate_rule({rewrite, Type, Source, Re, Dest}) ->
+    case validate_topic(filter, Source) of
+        {ok, _} ->
+            case validate_topic(dest_topic_type(Type), Dest) of
+                {ok, _} ->
+                    {ok, {rewrite, Source, Re, Dest}};
+                {error, Reason} ->
+                    {error, {dest, Dest}, Reason}
+            end;
+        {error, Reason} ->
+            {error, {source, Source}, Reason}
+    end.
+
+log_rules(Type, Rules) ->
+    log_rules_(Type, Rules, 1).
 
-log_rule([], _Index) -> ok;
-log_rule([{Type, Topic, Re, Dest} | Rules], Index) ->
+log_rules_(_Type, [], _Index) -> ok;
```

Comment:
```suggestion
do_log_rules(_Type, [], _Index) -> ok;
```

Reply:
or `log_rules_loop`
if i remember correctly, elvis does not like functions ending with `_`